crates/derive/src/online/validation.rs

```diff
@@ -0,0 +1,92 @@
+//! Contains logic to validate derivation pipeline outputs.
+
+use crate::types::{L2AttributesWithParent, L2PayloadAttributes, RawTransaction};
+use alloc::{boxed::Box, vec};
+use alloy_provider::{Provider, ReqwestProvider};
+use alloy_rpc_types::{Block, BlockNumberOrTag, BlockTransactions};
+use alloy_transport::TransportResult;
+use anyhow::Result;
+use async_trait::async_trait;
+
+/// Validator
+///
+/// The validator trait describes the interface for validating the derivation outputs.
+#[async_trait]
+pub trait Validator {
+    /// Validates the given [`L2AttributesWithParent`].
+    async fn validate(&self, attributes: &L2AttributesWithParent) -> bool;
+}
+
+/// OnlineValidator
+///
+/// Validates the [`L2AttributesWithParent`] by fetching the associated L2 block from
+/// a trusted L2 RPC and constructing the L2 Attributes from the block.
+#[derive(Debug, Clone)]
+pub struct OnlineValidator {
+    /// The L2 provider.
+    provider: ReqwestProvider,
+}
+
+impl OnlineValidator {
+    /// Creates a new `OnlineValidator`.
+    pub fn new(provider: ReqwestProvider) -> Self {
+        Self { provider }
+    }
+
+    /// Creates a new [OnlineValidator] from the provided [reqwest::Url].
+    pub fn new_http(url: reqwest::Url) -> Self {
+        let inner = ReqwestProvider::new_http(url);
+        Self::new(inner)
+    }
+
+    /// Fetches Transactions from the L2 provider.
+    pub(crate) async fn get_block(&self, tag: BlockNumberOrTag) -> Result<Block> {
+        let method = alloc::borrow::Cow::Borrowed("eth_getBlockByNumber");
+        let block: TransportResult<Block> = self.provider.raw_request(method, (tag, true)).await;
+        let block = block.map_err(|e| anyhow::anyhow!(e))?;
+        Ok(block)
+    }
+
+    /// Gets the payload for the specified [BlockNumberOrTag].
+    pub(crate) async fn get_payload(&self, tag: BlockNumberOrTag) -> Result<L2PayloadAttributes> {
+        // TODO: we can't use the provider's get_block_by_number here because
+        // upstream alloy will return Mainnet Ethereum transactions and not work for Optimism
+        // Deposit Transactions.
+        let block = self.get_block(tag).await?;
```

Comment:
This will not work for L2 blocks, since the `N = Ethereum` bound does not support OP transactions. Longer term I think we want to implement the `Network` trait for OP, though for now, can use `debug_*` endpoints to get the raw tx RLP and decode into `OpTxEnvelope`.